File: lachancuoicgoi_flutter/lib/analysis/l3/core/gemini_chat_session.dart
import 'dart:async';

import 'package:google_generative_ai/google_generative_ai.dart';

import 'api_key_provider.dart';
import 'gemini_client.dart';
import 'gemini_config.dart';
import 'gemini_metrics.dart';
import 'key_health_tracker.dart';
import 'operation_result.dart';

typedef GeminiChatExecutor =
    Future<String> Function({
      required String apiKey,
      required GeminiConfig config,
      required String modelName,
      required List<Content> history,
      required String prompt,
    });

class GeminiChatSession {
  GeminiChatSession({
    required this.apiKeyProvider,
    required this.config,
    this.keyHealthTracker,
    GeminiChatExecutor? chatExecutor,
  }) : _chatExecutor = chatExecutor ?? _defaultChatExecutor;

  static const Duration _minInterval = Duration(seconds: 1);
  // Max number of Content entries in _safeHistory (user + model = 2 per exchange).
  // 20 entries = 10 exchanges. Prevents exceeding Gemini context window.
  static const int _maxHistoryEntries = 20;
  static const List<String> _fallbackModels = geminiFallbackModels;

  final ApiKeyProvider apiKeyProvider;
  final GeminiConfig config;
  final KeyHealthTracker? keyHealthTracker;
  final GeminiChatExecutor _chatExecutor;

  int _currentKeyIndex = 0;
  int _currentModelIndex = 0;
  DateTime? _lastCallTime;
  // Rate-limit serialization chain (see _applyRateLimit). Prevents concurrent
  // sendMessage() calls from both reading a stale _lastCallTime and firing
  // closer together than _minInterval.
  Future<void> _rateLimitChain = Future<void>.value();
  final List<Content> _safeHistory = <Content>[];

  Future<Result<T>> sendMessage<T>(
    String text,
    T Function(String responseText, String modelName) parser,
  ) async {
    await _applyRateLimit();
    final startTime = DateTime.now();
    final keys = apiKeyProvider.getApiKeys();
    if (keys.isEmpty) {
      return Result.failure(StateError('No API keys available'));
    }

    final activeIndices = <int>[];
    final bestKey =
        keyHealthTracker?.getAvailableKeyIndex() ?? _currentKeyIndex;
    if (bestKey >= 0) {
      activeIndices.add(bestKey);
    }
    final others =
        keyHealthTracker?.getActiveKeyIndices() ??
        List<int>.generate(keys.length, (index) => index);
    for (final index in others) {
      if (!activeIndices.contains(index)) {
        activeIndices.add(index);
      }
    }
    if (activeIndices.isEmpty) {
      return Result.failure(
        StateError('All API keys are exhausted or in cooldown'),
      );
    }

    Object? lastError;
    StackTrace? lastStackTrace;
    var firstKeyAttempted = false;

    for (final keyIndex in activeIndices) {
      if (!firstKeyAttempted) {
        firstKeyAttempted = true;
      } else {
        await Future<void>.delayed(const Duration(seconds: 1));
      }

      if (keyIndex != _currentKeyIndex) {
        _currentKeyIndex = keyIndex;
        _currentModelIndex = 0;
      }

      for (
        var modelIndex = _currentModelIndex;
        modelIndex < _fallbackModels.length;
        modelIndex++
      ) {
        _currentModelIndex = modelIndex;
        final modelName = _fallbackModels[_currentModelIndex];

        int modelRetries = 0;
        bool shouldMoveToNextModel = false;

        while (modelRetries < 2 && !shouldMoveToNextModel) {
          if (modelRetries > 0) {
            await Future<void>.delayed(const Duration(milliseconds: 1000));
          }

          try {
            final responseText = await _chatExecutor(
              apiKey: keys[_currentKeyIndex],
              config: config,
              modelName: modelName,
              history: List<Content>.unmodifiable(_safeHistory),
              prompt: text,
            );
            final parsed = parser(responseText, modelName);

            // L3 Resilience: Cost tracking
            final historyTextLength = _safeHistory.fold<int>(
              0,
              (sum, content) =>
                  sum +
                  (content.parts.firstOrNull is TextPart
                      ? (content.parts.first as TextPart).text.length
                      : 0),
            );
            final estimatedTokens =
                (historyTextLength + text.length + responseText.length) ~/ 4;
            keyHealthTracker?.recordTokenUsage(
              _currentKeyIndex,
              estimatedTokens,
            );

            _safeHistory.add(Content.text(text));
            _safeHistory.add(Content.model(<Part>[TextPart(responseText)]));
            while (_safeHistory.length > _maxHistoryEntries) {
              _safeHistory.removeRange(0, 2);
            }
            keyHealthTracker?.markSuccess(_currentKeyIndex);
            GeminiMetrics.instance.recordCall(
              success: true,
              latencyMs: DateTime.now().difference(startTime).inMilliseconds,
              keyIndex: _currentKeyIndex,
            );
            return Result.success(parsed);
          } on Object catch (error, stackTrace) {
            lastError = error;
            lastStackTrace = stackTrace;
            final errorType = _classifyError(error);
            if (errorType == GeminiErrorType.auth) {
              keyHealthTracker?.markInvalid(_currentKeyIndex, error.toString());
              shouldMoveToNextModel = true;
              break;
            }
            if (errorType == GeminiErrorType.quota ||
                errorType == GeminiErrorType.modelNotFound) {
              shouldMoveToNextModel = true;
              continue;
            }
            if (error is TimeoutException) {
              modelRetries++;
              keyHealthTracker?.markError(_currentKeyIndex, error.toString());
              continue;
            }
            modelRetries++;
            keyHealthTracker?.markError(_currentKeyIndex, error.toString());
          }
        }

        if (_classifyError(lastError) == GeminiErrorType.auth) {
          break;
        }
      }

      if (_classifyError(lastError) == GeminiErrorType.quota) {
        keyHealthTracker?.markQuotaExceeded(_currentKeyIndex);
      }
    }

    return Result.failure(
      lastError ?? StateError('Unknown error during chat session fallback'),
      lastStackTrace,
    );
  }

  void close() {
    _safeHistory.clear();
    _currentModelIndex = 0;
  }

  Future<void> _applyRateLimit() async {
    // Serialize through a Future chain so concurrent callers queue up rather
    // than each reading a stale _lastCallTime and firing together.
    final completer = Completer<void>();
    final previous = _rateLimitChain;
    _rateLimitChain = completer.future;
    try {
      await previous;
      final lastCall = _lastCallTime;
      if (lastCall != null) {
        final elapsed = DateTime.now().difference(lastCall);
        if (elapsed < _minInterval) {
          await Future<void>.delayed(_minInterval - elapsed);
        }
      }
      _lastCallTime = DateTime.now();
    } finally {
      completer.complete();
    }
  }

  GeminiErrorType _classifyError(Object? error) => classifyGeminiError(error);

  static Future<String> _defaultChatExecutor({
    required String apiKey,
    required GeminiConfig config,
    required String modelName,
    required List<Content> history,
    required String prompt,
  }) async {
    final model = GenerativeModel(
      model: modelName,
      apiKey: apiKey,
      generationConfig: GenerationConfig(
        temperature: config.temperature,
        topK: config.topK,
        topP: config.topP,
        responseMimeType: config.responseMimeType,
      ),
    );
    final chat = model.startChat(history: history);
    final response = await chat
        .sendMessage(Content.text(prompt))
        .timeout(config.timeout);
    final text = response.text?.trim();
    if (text == null || text.isEmpty) {
      throw StateError('Empty response from Gemini');
    }
    return text;
  }
}

File: lachancuocgoi_kotline/app/src/main/java/com/example/lachancuocgoi/Analysis/L3/core/GeminiChatSession.kt
package com.example.lachancuocgoi.Analysis.L3.core

import android.util.Log
import com.example.lachancuocgoi.Analysis.AnalysisResult
import com.example.lachancuocgoi.RiskLevel
import com.google.ai.client.generativeai.Chat
import com.google.ai.client.generativeai.GenerativeModel
import com.google.ai.client.generativeai.type.generationConfig
import com.google.ai.client.generativeai.type.Content
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.delay
import kotlinx.coroutines.withTimeout

/**
 * Quản lý conversation session với Gemini API.
 * 
 * Features:
 * - Maintain single chat session cho mỗi cuộc gọi
 * - Auto-switch API keys với context preservation
 * - Rate limiting để tránh quota exhaustion
 * - Thread-safe operations với Mutex
 */
class GeminiChatSession(
    private val apiKeyProvider: ApiKeyProvider,
    private val config: GeminiConfig,
    private val keyHealthTracker: KeyHealthTracker? = null
) {
    companion object {
        private const val TAG = "GeminiChatSession"
        private const val MIN_INTERVAL_MS = 1000L  // Minimum 1s between calls
    }
    
    // Current active chat session
    private var currentChat: Chat? = null
    private var currentKeyIndex = 0
    private var currentModelIndex = 0
    private var hasNotifiedAllExhausted = false
    
    // Priority order for Gemini models
    private val fallbackModels = listOf(
        "gemini-2.5-flash-lite",
        "gemini-3-flash-preview",
        "gemini-2.5-flash",
        "gemini-3.1-flash-lite-preview"
    )

    // Track conversation for context preservation
    private val conversationHistory = mutableListOf<String>()
    
    // Lưu lịch sử an toàn của SDK để tránh lỗi Corrupted State
    private var safeHistory: List<Content> = emptyList()
    
    // Thread safety
    private val sessionLock = Mutex()
    
    // Rate limiting
    private var lastCallTime = 0L
    
    /**
     * Gửi message trong chat session hiện tại.
     * 
     * @param text Message content (prompt)
     * @param parser Function để parse response text thành result type
     * @return Result<T> chứa parsed value hoặc error
     */
    suspend fun <T> sendMessage(
        text: String,
        parser: (String, String) -> T
    ): Result<T> = sessionLock.withLock {
        // Rate limiting
        applyRateLimit()
        
        val startTime = System.currentTimeMillis()
        val keys = apiKeyProvider.getApiKeys()
        
        if (keys.isEmpty()) {
            return Result.failure(Exception("No API keys available"))
        }

        // Lấy danh sách key ưu tiên
        val activeIndices = mutableListOf<Int>()
        val bestKey = keyHealthTracker?.getAvailableKeyIndex()
        if (bestKey != null && bestKey >= 0) {
            activeIndices.add(bestKey)
        } else {
            activeIndices.add(currentKeyIndex)
        }
        val otherKeys = keyHealthTracker?.getActiveKeyIndices()?.filter { it != activeIndices[0] } 
            ?: (0 until keys.size).filter { it != activeIndices[0] }
        activeIndices.addAll(otherKeys)

        if (activeIndices.isEmpty() && keyHealthTracker != null) {
            return Result.failure(Exception("All API keys are exhausted or in cooldown"))
        }

        // Bắt đầu vòng lặp fallback (tương tự L3Analyzer)
        var lastError: Exception? = null
        var firstKeyAttempted = false

        for (keyIndex in activeIndices) {
            // Refresh timeout nếu phải chuyển API key
            if (activeIndices.size > 1 && !firstKeyAttempted) {
                firstKeyAttempted = true
            } else if (firstKeyAttempted) {
                delay(1000)
            }

            // Nếu đang dùng key mới so với session hiện tại, reset model index về 0
            if (keyIndex != currentKeyIndex) {
                currentKeyIndex = keyIndex
                currentModelIndex = 0
                currentChat = null
            }

            for (modelIndex in currentModelIndex until fallbackModels.size) {
                currentModelIndex = modelIndex
                val modelName = fallbackModels[currentModelIndex]

                // Create or restore chat
                if (currentChat == null) {
                    currentChat = createNewChat(currentKeyIndex, modelName, safeHistory)
                    Log.d(TAG, "Created/Restored chat session with key $currentKeyIndex, model $modelName")
                }

                // Lưu lịch sử an toàn trước khi gọi
                safeHistory = currentChat?.history?.filterNotNull() ?: safeHistory

                try {
                    val response = withTimeout(config.timeoutMs) {
                        applyRateLimit()
                        currentChat!!.sendMessage(text)
                    }

                    val responseText = response.text ?: throw Exception("Empty response from Gemini")
                    val parsed = parser(responseText, modelName)

                    // Track history and mark success
                    conversationHistory.add("User: ${text.take(50)}... | AI: ${responseText.take(50)}...")
                    safeHistory = currentChat?.history?.filterNotNull() ?: safeHistory
                    keyHealthTracker?.markSuccess(currentKeyIndex)

                    val latency = System.currentTimeMillis() - startTime
                    GeminiMetrics.recordCall(success = true, latencyMs = latency)
                    Log.d(TAG, "✓ Message sent successfully in ${latency}ms using $modelName on key $currentKeyIndex")

                    return Result.success(parsed)
                } catch (e: Exception) {
                    lastError = e
                    currentChat = null // Hủy session đang dính lỗi
                    
                    if (e is kotlinx.coroutines.CancellationException && e !is kotlinx.coroutines.TimeoutCancellationException) {
                        throw e
                    }

                    val msg = e.message?.lowercase() ?: ""
                    val isQuotaError = msg.contains("429") || msg.contains("quota")
                    val isAuthError = msg.contains("403") || msg.contains("api key") || msg.contains("api_key")

                    Log.w(TAG, "Error with key $currentKeyIndex, model $modelName: ${e.message}")

                    if (isAuthError) {
                        keyHealthTracker?.markInvalid(currentKeyIndex, e.message)
                        break // Bỏ qua các model khác, nhảy sang key mới
                    } else if (isQuotaError) {
                        // Thử model tiếp theo trên CÙNG 1 key
                        continue 
                    } else if (e is NullPointerException) {
                        Log.w(TAG, "NPE từ SDK. Thử model/key khác.")
                        keyHealthTracker?.markError(currentKeyIndex, e.message)
                        break
                    } else {
                        keyHealthTracker?.markError(currentKeyIndex, e.message)
                        Log.w(TAG, "Lỗi mạng/server (Không phải Quota). Abort.")
                        return Result.failure(e)
                    }
                }
            }

            // Nếu vòng lặp model kết thúc mà vẫn fail, tức là key này đã kiệt quệ => Qua key khác
            keyHealthTracker?.markQuotaExceeded(currentKeyIndex)
            Log.w(TAG, "All models failed on key $currentKeyIndex. Moving to next key.")
        }

        if (keyHealthTracker?.areAllKeysDown() == true) {
            Log.e(TAG, "ALL keys & models exhausted! Quota resets at 00:00.")
        }

        return Result.failure(lastError ?: Exception("Unknown error during chat session fallback"))
    }
    
    /**
     * Đóng chat session hiện tại.
     */
    fun close() {
        currentChat = null
        conversationHistory.clear()
        safeHistory = emptyList() // Clear an toàn
        Log.d(TAG, "Chat session closed")
    }
    
    /**
     * Tạo Chat instance mới với API key, model name, và khôi phục Context.
     */
    private fun createNewChat(keyIndex: Int, modelName: String, history: List<Content> = emptyList()): Chat {
        val keys = apiKeyProvider.getApiKeys()
        val apiKey = keys[keyIndex]
        
        val genConfig = generationConfig {
            temperature = config.temperature
            topK = config.topK
            topP = config.topP
            config.responseMimeType?.let { responseMimeType = it }
        }
        
        val model = GenerativeModel(
            modelName = modelName,
            apiKey = apiKey,
            generationConfig = genConfig
        )
        
        return model.startChat(history)
    }
    
    /**
     * Apply rate limiting để tránh quota exhaustion.
     */
    private suspend fun applyRateLimit() {
        val now = System.currentTimeMillis()
        val timeSinceLastCall = now - lastCallTime
        
        if (timeSinceLastCall < MIN_INTERVAL_MS) {
            val waitTime = MIN_INTERVAL_MS - timeSinceLastCall
            delay(waitTime)
        }
        
        lastCallTime = System.currentTimeMillis()
    }
}